eng/scripts/test-matrix/test-matrix-functions.ps1

```diff
@@ -0,0 +1,317 @@
+    $exclusionMatrix = @()
+
+    foreach ($exclusion in $excludes) {
+        $converted = ConvertToMatrixArrayFormat $exclusion
+        $full, $_ = GenerateFullMatrix $converted
+        $exclusionMatrix += $full
+    }
+
+    foreach ($element in $matrix) {
+        foreach ($exclusion in $exclusionMatrix) {
+            $match = MatrixElementMatch $element.parameters $exclusion.parameters
+            if ($match) {
+                $element.parameters[$deleteKey] = $true
+            }
+        }
+    }
+
+    return $matrix | Where-Object { !$_.parameters.Contains($deleteKey) }
+}
+
+function ProcessIncludes([Array]$matrix, [Array]$includes, [HashTable]$displayNames) {
+    foreach ($inclusion in $includes) {
+        $converted = ConvertToMatrixArrayFormat $inclusion
+        $full, $_ = GenerateFullMatrix $converted $displayNames
+        $matrix += $full
+    }
+
+    return $matrix
+}
+
+function MatrixElementMatch([HashTable]$source, [HashTable]$target) {
+    if ($target.Count -eq 0) {
+        return $false
+    }
+
+    foreach ($key in $target.Keys) {
+        if ($source[$key] -ne $target[$key]) {
+            return $false
+        }
+    }
+
+    return $true
+}
+
+function ConvertToMatrixArrayFormat([HashTable]$matrix) {
+    $converted = @{}
+
+    foreach ($key in $matrix.Keys) {
+        if ($matrix[$key] -isnot [Array]) {
+            $converted[$key] = ,$matrix[$key]
+        } else {
+            $converted[$key] = $matrix[$key]
+        }
+    }
+
+    return $converted
+}
+
+function SerializePipelineMatrix([Array]$matrix) {
+    $matrix = $matrix | Sort-Object -Property name
+    $pipelineMatrix = [ordered]@{}
+    foreach ($entry in $matrix) {
+        $pipelineMatrix.Add($entry.name, [ordered]@{})
+        foreach ($key in $entry.parameters.Keys) {
+            $pipelineMatrix[$entry.name].Add($key, $entry.parameters[$key])
+        }
+    }
+
+    return @{
+        compressed = $pipelineMatrix | ConvertTo-Json -Compress ;
+        pretty = $pipelineMatrix | ConvertTo-Json;
+    }
+}
+
+function GenerateSparseMatrix([HashTable]$parameters, [HashTable]$displayNames, [int]$count) {
+    [Array]$matrix, [Array]$dimensions = GenerateFullMatrix $parameters $displayNames
+    $sparseMatrix = @()
+
+    # With full matrix, retrieve items by doing diagonal lookups across the matrix N times.
+    # For example, given a matrix with dimensions 3, 2, 2:
+    # 0, 0, 0
+    # 1, 1, 1
+    # 2, 2, 2
+    # 3, 0, 0 <- 3, 3, 3 wraps to 3, 0, 0 given the dimensions
+    for ($i = 0; $i -lt $count; $i++) {
+        $idx = @()
+        for ($j = 0; $j -lt $dimensions.Length; $j++) {
+            $idx += $i % $dimensions[$j]
+        }
+        $sparseMatrix += GetNdMatrixElement $idx $matrix $dimensions
+    }
+
+    return $sparseMatrix
+}
+
+function GenerateFullMatrix([HashTable]$parameters, [HashTable]$displayNames = @{}) {
+    $sortedParameters = SortMatrix $parameters
+
+    $matrix = [System.Collections.ArrayList]::new()
```

Comment:
Why explicitly an ArrayList here?

Reply:
I need to expand the matrix array dynamically through the recursive function `InitializeMatrix`. If I pass an array down, and expand it via `array += <value>` it does not work by reference, whereas `ArrayList.Add()` does. The algorithm to generate all permutations of an Nd array is *much* simpler when there's a side effect.